### PR TITLE
[bugfix] avoid attention padding tokens computation in pcg

### PR DESCRIPTION
**Pull request by Chen-0210**
## Motivation

When PCG is enabled, the attention metadata is initialized with real_num_tokens, but the input tensor still contains padded tokens. Attention backends such as FlashInfer can not handle this well, which can lead to undefined behavior, including nan value, corrupted outputs (repeated !!!!!), result in abnormally long output lengths.

To fix this, exclude the padded tokens and make PCG more robust.

## Modifications
1.  Reverts  PR #21452
2. Slices tensors to real_num_tokens before/after the attention forward call — in `unified_attention_with_output` and `unified_linear_attention_with_output`, `query/key/value and out_cache_loc` are narrowed to exclude padding tokens before calling the backend, then restored out_cache_loc afterward, so attention only ever sees real tokens.

## Accuracy Tests
Tested in CI
## Checklist

- [x] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [x] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [x] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).

## Review Process

1. Ping Merge Oncalls to start the PR flow. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process).
3. Get approvals from [CODEOWNERS](https://github.com/sgl-project/sglang/blob/main/.github/CODEOWNERS) and other reviewers.
4. Trigger CI tests with [comments](https://docs.sglang.io/developer_guide/contribution_guide.html#how-to-trigger-ci-tests) or contact authorized users to do so.
   - `/tag-run-ci-label`, `/rerun-failed-ci`, `/tag-and-rerun-ci`
5. After green CI and required approvals, ask Merge Oncalls to merge.

### CODE REVIEW

File: python/sglang/srt/models/qwen3_next.py [hunk 1]
```diff
@@ -393,7 +393,7 @@ def forward(
         forward_batch: ForwardBatch,
     ):
         if forward_batch.forward_mode.is_extend() and get_forward_context() is not None:
-            output = torch.empty_like(hidden_states)
+            output = torch.zeros_like(hidden_states)
```
Comment (hebiao064) [inline python/sglang/srt/models/qwen3_next.py]: is this change necessary? i think the output will be overridden anyway?

File: python/sglang/srt/layers/radix_attention.py [hunk 2]
```diff
@@ -113,9 +113,9 @@ def forward(
 
         if forward_batch.forward_mode.is_extend() and get_forward_context() is not None:
             if self.qk_head_dim != self.v_head_dim:
-                output = q.new_empty((q.shape[0], self.tp_q_head_num * self.v_head_dim))
+                output = q.new_zeros((q.shape[0], self.tp_q_head_num * self.v_head_dim))
```
Comment (Oasis-Git) [inline python/sglang/srt/layers/radix_attention.py]: Why we need to set zero for the buffer? If required, maybe we should set it for linear model or next model and keep empty for normal case.

File: python/sglang/srt/model_executor/piecewise_cuda_graph_runner.py [hunk 3]
```diff
@@ -350,6 +301,7 @@ def warmup_torch_compile(self, num_tokens: int):
                 num_token_non_padded=None,
                 global_forward_mode=ForwardMode.EXTEND,
                 lora_ids=None,
+                real_num_tokens=num_tokens,
```
Comment (ch-wan) [inline python/sglang/srt/model_executor/piecewise_cuda_graph_runner.py]: we can use `num_token_non_padded`

File: python/sglang/srt/model_executor/forward_batch_info.py [hunk 4]
```diff
@@ -375,6 +375,9 @@ class ForwardBatch(ForwardBatchDeepSeekMHAMixin):
     # For hidden states before normal
     return_hidden_states_before_norm: bool = False
 
+    # For piecewise cuda graph
+    real_num_tokens: Optional[int] = None
```
Comment (ch-wan) [inline python/sglang/srt/model_executor/forward_batch_info.py]: it is equivalent to `num_token_non_padded`

**Review by Oasis-Git — COMMENTED** [review]
Since these changes affect a broad range of functionality, the full unit test suite should be carefully validated before approving this PR.

File: python/sglang/srt/layers/radix_attention.py (hunk 2 above)
Comment (Oasis-Git) [reply python/sglang/srt/layers/radix_attention.py]: For example, for deepseek v2 `out_cache_loc` might be used within cuda graph?

File: python/sglang/srt/model_executor/forward_batch_info.py (hunk 4 above)
Comment (Chen-0210) [reply python/sglang/srt/model_executor/forward_batch_info.py]: great! replace it with `num_token_non_padded_cpu`.

File: python/sglang/srt/layers/radix_attention.py (hunk 2 above)
Comment (Chen-0210) [reply python/sglang/srt/layers/radix_attention.py]: zeros is not a must, removed it

Comment (Chen-0210) [reply python/sglang/srt/layers/radix_attention.py]: > For example, for deepseek v2 `out_cache_loc` might be used within cuda graph?

What does this mean?

File: python/sglang/srt/models/qwen3_next.py (hunk 1 above)
Comment (Chen-0210) [reply python/sglang/srt/models/qwen3_next.py]: only the non-padding part is overwritten. But it will not have and side effect, drop it

File: python/sglang/srt/layers/radix_attention.py (hunk 2 above)
Comment (Oasis-Git) [reply python/sglang/srt/layers/radix_attention.py]: https://github.com/sgl-project/sglang/blob/81449b4beebf761eab1ad67ad19e3cab4c0109d4/python/sglang/srt/models/deepseek_v2.py#L1777.

For ds v2 model, out_cache_loc will be included in pcg.

Comment (Chen-0210) [reply python/sglang/srt/layers/radix_attention.py]: @Oasis-Git  I see, add out_cache_loc back

File: python/sglang/srt/model_executor/piecewise_cuda_graph_runner.py [hunk 5]
```diff
@@ -198,21 +190,6 @@ def __init__(self, model_runner: ModelRunner):
                 if model_runner.is_hybrid_swa
                 else None
             )
-            self.mamba_track_indices = (
```
Comment (hebiao064) [inline python/sglang/srt/model_executor/piecewise_cuda_graph_runner.py]: why we need to remove these code?

Comment (Chen-0210) [reply python/sglang/srt/model_executor/piecewise_cuda_graph_runner.py]: The relevant part is not included in PCG, so it does not require a fixed memory address.

**Comment by zminglei** [comment]
> image
> previously disabled CI `test_qwen3_next_models_pcg` is now enabled and pass.

Seems previously this CI only has a fixed chance to fail, not fail every time. Could we run the test several times, let's say 10 times to verify it passes every time?

**Comment by Chen-0210** [comment]
@zminglei It seems the re-run didn’t take effect

File: python/sglang/srt/layers/radix_attention.py [hunk 6]
```diff
@@ -152,21 +152,23 @@ def unified_attention_with_output(
...
     kwargs = {}
     if q_rope is not None:
-        kwargs["q_rope"] = q_rope
+        kwargs["q_rope"] = q_rope[:real_num_tokens, :]
     if k_rope is not None:
-        kwargs["k_rope"] = k_rope
+        kwargs["k_rope"] = k_rope[:real_num_tokens, :]
     if sinks is not None:
-        kwargs["sinks"] = sinks
+        kwargs["sinks"] = sinks[:real_num_tokens, :]
```
Comment (ispobock) [inline python/sglang/srt/layers/radix_attention.py]: Could you check the shape of attention sinks for models like gpt-oss and mimo? There is no token dim.

Comment (Chen-0210) [reply python/sglang/srt/layers/radix_attention.py]: Yes, see that, updated to keep sinks unchanged.

**Comment by Oasis-Git** [comment]
@Chen-0210 Could you fix the ci of stage-a?

**Review by Oasis-Git — APPROVED** [review]
I think this is a clean change. approved

**Comment by ispobock** [comment]
@Chen-0210 could you fix the test conflict?

**Comment by Chen-0210** [comment]
> @Chen-0210 could you fix the test conflict?

Done
